src/test/ui/async-await/unused-lifetime.rs

```diff
@@ -0,0 +1,42 @@
+// edition:2018
+
+// Avoid spurios warnings of unused lifetime. The below async functions
+// are desugered to have an unused lifetime
+// but we don't want to warn about that as there's nothing they can do about it.
+
+#![deny(unused_lifetimes)]
+#![allow(dead_code)]
+
+pub async fn october(s: &str) {
+    println!("{}", s);
+}
+
+pub async fn async_fn(&mut ref s: &mut[i32]) {
+    println!("{:?}", s);
+}
+
+macro_rules! foo_macro {
+    () => {
+        pub async fn async_fn_in_macro(&mut ref _s: &mut[i32]) {}
+    };
+}
+
+foo_macro!();
+
+pub async fn func_with_unused_lifetime<'a>(s: &'a str) {
```

Comment:
These shouldn't be linting.

Reply:
I thought that's cool, but am now wondering about setting self.lifetime_uses with LifetimeUseSet::Many. I can see it would prevent the lint from triggering, but is it strictly 'true'? It's my understanding that those autogenerated async lifetimes _do_ only have a single use. I worry that we might thwart current/later attempts at optimisations by setting the usage to Many. (Principle of least suprise)

(if an autogenerated async lifetime is actually used in mutliple locations then all my worries melt away - keen to here your thoughts @matthewjasper )

Reply:
I don't get why something like this should not be triggering the lint:

    pub async fn func_with_unused_lifetime<'a>(s: &'a str) {

The lint can hint that we can simplify it to:

    pub async fn func_with_unused_lifetime(s: &str) {

Or am I missing something? Ideally we want to get this closed out before tomorrow's beta branch. (ping @nikomatsakis )

Reply:
That should trigger `single_use_lifetimes` (it doesn't currently, but that's a different bug)

Reply:
> The lifetimes added here should have their IDs added to self.lifetime_uses with LifetimeUseSet::Many.

This seems to be too aggressive, when I do this the lint doesn't trigger in an async function unless we add in the check of name == kw::UnderscoreLifetime before adding LifetimeUseSet::Many.

Reply:
If I don't add to the inner else, it passes all the specific tests. @matthewjasper please have a look if these changes make sense to you. I couldn't think of a way to make getting the def_id any prettier short of doing the hashmap lookup again.

Reply:
I think this is resolved - let me know if not.